Refactor formula, cask and Ruby source downloads to use shared code

Bo98 opened the pull request:
This is significant refactor that makes formula resources (including bottles, source archives, resource archives and patches), cask downloads, formula Ruby source downloads (which previously was a resource), and cask Ruby source downloads all use a common `Downloadable` class.

It means everything that downloads should be more aligned, and easier to add new things to that system should it be necessary in the future.

This has a significant UX improvement for cask Ruby source downloads which previously had zero progress feedback and had no caching between brew runs.

The diff is large and this change probably took too long for most people (it was a fair bit of work while realistically being low priority) but here's an overview:
* Note the changes to `api/cask.rb` and `api/formula.rb` which have methods that download the Ruby source and load the file. They are effectively identical to each other beyond the usual formula & cask differences.
* `api/download.rb` shows how all that's on top of the `Downloadable` system is a nice constructor with nothing extra.
* Formulary no longer defines a resource - it now reads two parts of the JSON in the same way that we read the rest of it so there's nothing magic there anymore.
* `formula_installer` and `cask/installer` code to load the source is now a single line of code.
* `resource.rb` and `cask/download.rb` also use `Downloadable`
* There's a new `URL` class, though this is really only used internally to `Downloadable` and its subclasses at the moment.
* Checksum validation code duplication is no more.
* Cask source mirror code now uses existing mirror logic that formulae etc use.

Potential breaking change:
* `Resource#version=` has been removed (we had two setters before which seemed odd to me), but this API should be private AFAICT.

Closes #14746.

MikeMcQuaid commented on Library/Homebrew/api/cask.rb:
```diff
@@ -20,12 +21,22 @@ def fetch(token)
-        }
-        def fetch_source(token, path:, git_head:, sha256: nil)
-          Homebrew::API.fetch_homebrew_cask_source token, path: path, git_head: git_head, sha256: sha256
+        sig { params(cask: ::Cask::Cask).returns(::Cask::Cask) }
```
are all the `::` necessary? It feels like there's way more used in this PR than I typically see used?

Bo98 replied:
Will double check this - it is for `::URL` in `cask/download.rb`.

MikeMcQuaid replied:
Cool, thanks 🙇🏻

Bo98 replied:
This picks up `API::Cask` so is needed.

MikeMcQuaid commented on Library/Homebrew/cask/download.rb:
```diff
@@ -9,71 +10,75 @@ module Cask
+    end
+
+    sig {
+      override
```
What's this do/for?

Bo98 replied:
If the upstream type signature changes then this will error so that we make sure changes are reflected in subclasses too.

MikeMcQuaid replied:
Nice, gotcha!

MikeMcQuaid commented on Library/Homebrew/downloadable.rb:
```diff
@@ -0,0 +1,132 @@
+
+  include Context
+
+  sig { returns(T.nilable(URL)) }
+  attr_reader :url
```
When would a URL be `nil`?

Bo98 replied:
Ideally it wouldn't but it can be due to the way that `url` may not be set immediately.

For resources: URL is defined later in a DSL in a `do block` but the nature of that means it may not actually be called and could be nil - we had various `url.nil?` checks already in `resource.rb` prior to this PR.
For cask downloads: we have a concept of lazily loaded URLs where the URL can only be calculated at install time. Plus casks allow basically anything to be nil at the moment.

I spent a fair bit of time fighting that but I can try again, potentially by ditching the instance variable altogether and deferring a getter to subclasses.

MikeMcQuaid replied:
No, all good, leave it as-is, all makes sense and just curious 👍🏻

MikeMcQuaid commented on Library/Homebrew/downloadable.rb:
```diff
@@ -0,0 +1,132 @@
+
+  sig { void }
+  def initialize
+    @mirrors = T.let([], T::Array[String])
```
Is the `T.let` needed here?

MikeMcQuaid replied:
Cool. Like the `::`: if it's not needed, remove, if it is: resolve this and ignore me 😁

MikeMcQuaid commented on Library/Homebrew/downloadable.rb:
```diff
@@ -0,0 +1,132 @@
+  def downloader
+    @downloader ||= begin
+      primary_url, *mirrors = determine_url_mirrors
+      raise ArgumentError, "attempted to use a Downloadable without a URL!" if primary_url.blank?
```
Feels like the type system should catch this instead, ideally?

MikeMcQuaid commented on Library/Homebrew/resource.rb:
```diff
@@ -232,24 +179,28 @@ def sha256(val)
 
-    @version = detect_version(val)
+    @version = case val
+    when String  then Version.create(val)
+    when Version then val
+    else
+      raise TypeError, "version '#{val.inspect}' should be a string"
+    end
```
Could this be handled by Sorbet instead?

Bo98 replied:
Probably but it wouldn't be enforced in homebrew-core since we don't typecheck there yet.

MikeMcQuaid replied:
Would `brew readall` be sufficient to catch this? If so, might be nice to remove something like this and make `brew readall` always do Sorbet runtime type checking? If it isn't/wouldn't be enough: at most add `TODO` to investigate later.

Increasingly wondering if we've hit the time to make `HOMEBREW_DEVELOPER` also enable type checking at runtime, too. I've had `HOMEBREW_SORBET_RUNTIME` enabled for a while and everything's working fine. CC @dduugg @reitermarkus for people who might have opinions about this.

Bo98 replied:
> Would `brew readall` be sufficient to catch this?

If you mean `brew readall` with Sorbet runtime checking then yes it might do - though we should test this with an intentionally bad formula.

> Increasingly wondering if we've hit the time to make `HOMEBREW_DEVELOPER` also enable type checking at runtime, too. I've had `HOMEBREW_SORBET_RUNTIME` enabled for a while and everything's working fine. 

I'm fine with this. Do you notice any measurable overhead out of interest? The no-checking code is a little hacky so in the long term it might be an option to not bother with that.

dduugg replied:
I also have `HOMEBREW_SORBET_RUNTIME` enabled without issue (after some initial hiccups that I PR'd). I can see if Stripe folks have perf stats that they're willing to share. I vaguely recall a 0%-7% impact across services in the early days, depending on factors like whether there was a hot code path or deeply nested types that had to be unpacked. (IIRC Sorbet also skips some complex runtime validations as a result now, so don't be afraid to write complex types based on that anecdata.)

As a starting point, I suggest enabling logging of type errors rather than the default of raising them. See the Sorbet [docs](https://sorbet.org/docs/tconfiguration) and our current [override](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/standalone/sorbet.rb) code.

MikeMcQuaid replied:
@dduugg Honestly I think logging is going to be more annoying to end-users. If we do this I think we should just enable them for developers/CI and make them errors.

Bo98 replied:
Unless there's strong opposition and people want it gone now: I'll leave a comment saying this can go if typechecking is enabled for taps.

MikeMcQuaid commented on Library/Homebrew/resource.rb:
```diff
@@ -261,6 +212,14 @@ def patch(strip = :p1, src = nil, &block)
   end
 
+  def using
+    @url&.using
```
Same here: when would we want a resource with a `nil` URL? Could we lean on the type system for that?

MikeMcQuaid reviewed (commented):
Great work so far @Bo98! Really liking how this API is looking. Some general thoughts:
- would be good to have some basic direct tests for `downloadable` and/or ensure the existing indirect tests have 100% coverage
- would be good to avoid `::Class` except in cases where there are errors
- a few cases where moving code around in a post-Sorbet world: we could lean more into the type system to avoid bad types and `nil`s
- is there any code shared with `DownloadStrategy` here or is it just from `Resource`?
- it might make sense to consider splitting this into multiple PRs if anyone else is finding this a bit hard to review. Alternatively, if you'd rather land this once it's ✅: I'd suggest doing so when you'll be around for a while (both that day and over the next few days) to pick up an unexpected 💥 from the large refactoring.

Again: great work!

Bo98 commented on Library/Homebrew/cask/cask_loader.rb:
```diff
@@ -69,11 +69,16 @@ def self.can_load?(ref)
       attr_reader :token, :path
 
-      def initialize(path) # rubocop:disable Lint/MissingSuper
+      def initialize(path, tap: nil) # rubocop:disable Lint/MissingSuper
```
The parent class does not accept nil.

A better option might be to create a base class for both - not quite sure yet!

MikeMcQuaid replied:
Definitely not blocking for this PR.

```suggestion
      # Need to pass nil to parent class super which it doesn't support.
      # TODO: create a base class for both.
      # rubocop:disable Lint/MissingSuper
      def initialize(path, tap: nil) 
```

and the `# rubocop:enable Lint/MissingSuper` below would be more than enough for now. Thanks!

Bo98 replied:
I've created a base class for this.

Bo98 commented on Library/Homebrew/cask/cask_loader.rb:
```diff
@@ -69,11 +69,16 @@ def self.can_load?(ref)
         @token = path.basename(path.extname).to_s
+        # For files we've downloaded, they will be prefixed with `{URL MD5}--`.
+        # Remove that prefix to get the original filename.
+        @token = @token.split("--", 2).last if path.dirname == HOMEBREW_CACHE/"downloads"
```
Perhaps given it's in formulary as well, though I'm not sure where beyond a util method?

MikeMcQuaid replied:
@Bo98 I guess it's `DownloadStrategy` that decides this logic? a `DownloadStrategy.download_split_something` method (obviously not that name and maybe not even that class) feels like it'd most make sense?

reitermarkus replied:
`DownloadStrategy#resolved_basename` should return this.

Bo98 replied:
> `DownloadStrategy#resolved_basename` should return this.

It does, but is tricky to pass that result when it gets to formulary where it may reload the file in subprocesses - at which point the concept of a download strategy instance is long gone. So the logic really only works when inside the loaders themselves.

Bo98 replied:
Ok I've overhauled this by making an `api-source` cache folder which contains tap information and each file is symlinked as its original basename.

Does require some custom cleanup.rb logic, but relying on the original logic was incorrect anyway.

apainintheneck commented on Library/Homebrew/cask/url.rb:
```diff
@@ -1,46 +1,174 @@
-  extend T::Sig
-
+module Cask
+  # Class corresponding to the `url` stanza.
+  #
   # @api private
-  class DSL
+  class URL < Delegator
```
I wanted to do this eventually too.

apainintheneck commented on Library/Homebrew/software_spec.rb:
```diff
@@ -91,15 +91,17 @@ def owner=(owner)
-      r.version ||= begin
+      next if r.version
+
+      r.version(begin
         raise "#{full_name}: version missing for \"#{r.name}\" resource!" if version.nil?
 
         if version.head?
           Version.create("HEAD")
         else
           version.dup
         end
-      end
+      end)
```
Maybe something like this would make more sense while you're in this file.

```rb
raise "#{full_name}: version missing for \"#{r.name}\" resource!" if version.nil?

r.version(version.head? ? Version.create("HEAD") : version.dup)
```

apainintheneck reviewed (commented):
In general, it looks really good. Not sure if it's already been mentioned here but it'd be nice to have some specs for the `Downloadable` class though.

apainintheneck commented on Library/Homebrew/formula.rb:
```diff
@@ -2081,6 +2081,18 @@ def missing_dependencies(hide: nil)
   end
 
+  # @private
+  sig { returns(T.nilable(String)) }
+  def ruby_source_path
+    path.relative_path_from(tap.path).to_s if path.exist?
+  end
+
+  # @private
+  sig { returns(T.nilable(Checksum)) }
+  def ruby_source_checksum
+    Checksum.new(Digest::SHA256.file(path).hexdigest) if path.exist?
+  end
```
```suggestion
  sig { returns(T.nilable(String)) }
  def ruby_source_path
    path.relative_path_from(tap.path).to_s if path&.exist?
  end

  # @private
  sig { returns(T.nilable(Checksum)) }
  def ruby_source_checksum
    Checksum.new(Digest::SHA256.file(path).hexdigest) if path&.exist?
  end
```

It seems like `path` will end up being `nil` if you load it from a source download.

```rb
[13] brew(main)> jrnl = "jrnl".f; nil
=> nil
[14] brew(main)> formula = Homebrew::API::Formula.source_download(jrnl); nil
==> Downloading https://raw.githubusercontent.com/Homebrew/homebrew-core/f6b1d5d773a1c8a7ddffcccb585
Already downloaded: /Users/kevinrobell/Library/Caches/Homebrew/downloads/a849e1f4e3269b75e55c85d56e9840553a665117d667d24dc53f62755455eed2--jrnl.rb
=> nil
[15] brew(main)> formula.to_hash
NoMethodError: undefined method `path' for nil:NilClass
from /usr/local/Homebrew/Library/Homebrew/formula.rb:2087:in `ruby_source_path'
```

This is just an edge case though that users would probably never run into.

MikeMcQuaid replied:
> This is just an edge case though that users would probably never run into.

IMO it's still worth defending against given that.

MikeMcQuaid replied:
This also seems like something's gone awry with type-checking if that can happen as Sorbet would normally warn about that.

Bo98 replied:
This was actually to do with a nil `tap` - not a nil `path` - and is already resolved in #15291.

Bo98 commented:
I've moved the cached downloads from the base of `HOMEBREW_CACHE` to `HOMEBREW_CACHE/api-source`. I realised that this is more correct to do - we split formula and cask downloads already and cleanup.rb largely assumed the base of `HOMEBREW_CACHE` is formula related downloads.

MikeMcQuaid reviewed (commented):
I'm good with this being merged when:
- CI is 🟢 
- there's even one unit test in a `downloadable_spec.rb` file
- `brew tests --coverage` or CodeCov indicates there's reasonable coverage of `downloadable.rb` to warrant not adding more tests.

Bo98 commented:
> there's even one unit test in a `downloadable_spec.rb` file

I thought about this but basically `Downloadable` is not designed to be used directly. It's effectively an abstract class that should be subclassed in the places that need it.

Coverage appears to be 95.95%. Not sure what the 5.05% is, because the file viewer is still broken:

<img width="1045" alt="Screenshot 2023-04-27 at 14 08 42" src="https://user-images.githubusercontent.com/1190754/234872253-52a80c37-17c5-4a14-bdfd-5ba017613988.png">

Bo98 commented:
Should work now: https://app.codecov.io/gh/Homebrew/brew/pull/15257/blob/Library/Homebrew/downloadable.rb

MikeMcQuaid commented:
Thanks again @Bo98!
